src/main/java/org/springframework/integration/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -452,13 +445,33 @@ public UpdateLeadersTask(Iterable<Partition> partitionsToReset) {
 
 			@Override
 			public void run() {
-				FastList<Partition> partitionsAsList = FastList.newList(partitionsToReset);
-				FastList<String> topics = partitionsAsList.collect(new PartitionToTopicFunction()).distinct();
-				kafkaTemplate.getConnectionFactory().refreshMetadata(topics);
-				Map<Partition, BrokerAddress> leaders = kafkaTemplate.getConnectionFactory().getLeaders(partitionsToReset);
-				synchronized (partitionsByBrokerMap) {
-					forEachKeyValue(leaders, new AddPartitionToBrokerProcedure());
-					partitionsByBrokerMap.notifyAll();
+				// fetch can complete successfully or unsuccessfully
+				boolean fetchCompleted = false;
+				while (!fetchCompleted && isRunning()) {
+					try {
+						FastList<Partition> partitionsAsList = FastList.newList(partitionsToReset);
+						FastList<String> topics = partitionsAsList.collect(new PartitionToTopicFunction()).distinct();
+						kafkaTemplate.getConnectionFactory().refreshMetadata(topics);
+						Map<Partition, BrokerAddress> leaders = kafkaTemplate.getConnectionFactory().getLeaders(partitionsToReset);
+						synchronized (partitionsByBrokerMap) {
+							forEachKeyValue(leaders, new AddPartitionToBrokerProcedure());
+							partitionsByBrokerMap.notifyAll();
+						}
+						System.out.println("Partitions fetched successfully");
+						fetchCompleted = true;
+					}
+					catch (Exception e) {
+						if (isRunning()) {
+							try {
+								Thread.sleep(DEFAULT_WAIT_FOR_LEADER_REFRESH_RETRY);
+							}
+							catch (InterruptedException e1) {
+								Thread.currentThread().interrupt();
+								log.error("Interrupted after refresh leaders failure");
+								fetchCompleted = true;
```

Comment:
Actually, that's not what we want. We want the container to continue running and resume operation when the target broker has come back alive.